src/hnsw/validate_index.c

```diff
@@ -0,0 +1,650 @@
+#include <postgres.h>
+
+#include "hnsw/validate_index.h"
+
+#include <access/heapam.h>  /* relation_open */
+#include <assert.h>         /* assert */
+#include <catalog/index.h>  /* IndexGetRelation */
+#include <inttypes.h>       /* PRIu32 */
+#include <stdint.h>         /* UINT32_MAX */
+#include <string.h>         /* bzero */
+#include <utils/memutils.h> /* AllocSetContextCreate */
+
+#include "hnsw/external_index.h" /* HnswIndexHeaderPage */
+#include "hnsw/options.h"        /* ldb_HnswGetM */
+
+enum ldb_vi_block_type
+{
+    LDB_VI_BLOCK_UNKNOWN,
+    LDB_VI_BLOCK_HEADER,
+    LDB_VI_BLOCK_BLOCKMAP,
+    LDB_VI_BLOCK_NODES,
+    LDB_VI_BLOCK_NR,
+};
+
+/* represents PostgreSQL block in the index */
+struct ldb_vi_block
+{
+    enum ldb_vi_block_type vp_type;
+    uint32_t               vp_nodes_nr;
+};
+
+/* represents a stored usearch node */
+struct ldb_vi_node
```

Comment:
note: assumes internal state about usearch. 
COuld you write a comment about this?
We will soon move to usearch v3 and may need to come update this

Reply:
Added a note.